cheatsheets/Authentication_Cheat_Sheet.md

```diff
@@ -110,24 +110,87 @@ Incorrectly implemented error messages in the case of authentication functionali
+* The user ID or password was incorrect.
+* The account does not exist.
+* The account is locked or disabled.
 
-#### Incorrect Response Examples
+The account creation feature is also impacted and the same approach of generic error message can be applied regarding the case in which the user exists.
 
+The objective is to prevent to create a [discrepancy factor](https://cwe.mitre.org/data/definitions/204.html) allowing an attacker to mount a user enumeration action against the feature.
+
+It is interesting to note that the business logic itself can bring a discrepancy factor related to the processing time taken. Indeed, depending on the implementation, the processing time can be significantly different according to the case (success vs failure) allowing an attacker to mount a [time-based attack](https://en.wikipedia.org/wiki/Timing_attack) (delta of some seconds for example).
+
+Example using pseudo-code for a login feature:
+
+*First implementation using the "quick exit" approach*
+
+```
+IF USER_EXISTS(login) THEN
+    password_hash=HASH(password)
+    IS_VALID=LOOKUP_CREDENTIALS_IN_STORE(login, password_hash)
+    IF NOT IS_VALID THEN
+        RETURN Error("Invalid User or Password!")    
+    ENDIF
+ELSE
+   RETURN Error("Invalid User or Password!")
+ENDIF
+```
+
+With this implementation: We can have a time-based discrepancy factor because, if the user exists then two time-taking operations are performed otherwise only one if the user does not exist.
+
+*Second implementation do not using the "quick exit" approach:*
+
+```
+password_hash=HASH(password)
+IS_VALID=LOOKUP_CREDENTIALS_IN_STORE(login, password_hash)
+IF NOT IS_VALID THEN
+   RETURN Error("Invalid User or Password!")
+ENDIF
+```
+
+With this implementation: The presence of a time-based discrepancy factor is less possible because, whatever if the user exists or not then two time-taking operations are always performed.
+
+The main issue with the fact to return a generic error message is that it will decrease the *user experience* of the feature and can confuse a legitimate user because the generic error message will not allow him to understand the source of the issue preventing him to access to his account/data. The decision to return or not a *generic error message* can be determined according to the criticity of the application and the data managed. For example, for critical applications, we can decide that in case of trouble, a user will be redirected to the helpdesk and a *generic error message* will always be returned.
+
+Regarding the user enumeration itself, protection against [brute-force attack](Authentication_Cheat_Sheet.md#prevent-brute-force-attacks) are also effective because they prevent an attacker to apply the enumeration at scale. Usage of [CAPTCHA](https://en.wikipedia.org/wiki/CAPTCHA) can be applied on a feature for which a *generic error message* cannot be returned because the *user experience* must be preserved.
```

Comment:
Do you think it would be nice to note the risk and acceptance level that we think of as security folks? Like if we were to return a specific message, then we need to cater for rate limiting and brute-force attacks, and how we model it? Or it would feel too much for the cheat sheet?

Reply:
To discuss refering to our chat on Slack before to add the content.